pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go

```diff
@@ -351,36 +380,49 @@ func (a *index) AddressInformation(addresses sets.String) ([]model.AddressInfo,
 	return res, sets.New(removed...)
 }
 
-func (a *index) WorkloadsForWaypoint(scope model.WaypointScope) []model.WorkloadInfo {
-	// Lookup scope. If its namespace wide, remove entries that are in SA scope
-	workloads := a.workloads.ByOwningWaypoint.Lookup(scope)
-	workloads = model.SortWorkloadsByCreationTime(workloads)
-	if scope.ServiceAccount == "" {
-		// This is a namespace wide waypoint. Per-SA waypoints have precedence, so we need to filter them out
-		workloads = slices.FilterInPlace(workloads, func(info model.WorkloadInfo) bool {
-			s := model.WaypointScope{
-				Namespace:      info.Namespace,
-				ServiceAccount: info.ServiceAccount,
-			}
-			// If there is a more specific waypoint, skip this
-			return len(a.waypoints.ByScope.Lookup(s)) == 0
-		})
+func (a *index) WorkloadsForWaypoint(key model.WaypointKey) []model.WorkloadInfo {
+	// TODO: we should be able to handle multiple IPs or a hostname
+	if len(key.Addresses) == 0 {
+		return nil
+	}
+	workloads := a.workloads.ByOwningWaypoint.Lookup(networkAddress{
+		network: key.Network,
+		ip:      key.Addresses[0],
+	})
+	services := a.services.ByOwningWaypoint.Lookup(networkAddress{
+		network: key.Network,
+		ip:      key.Addresses[0],
+	})
+	workloadsFromServices := make([]model.WorkloadInfo, 0)
```

Comment:
I don't think we need to get the workloads from the services here unless we're still supporting some "workload targeted traffic for a waypoint-captured Service.

Reply:
This is a great catch! The state now is that envoy gets a superset of the config it strictly requires but ztunnel config looks correct. I propose we proceed with the PR more or less as it sits now and address in a fast follower per https://github.com/istio/istio/issues/50005, if that's OK with you.

Reply:
https://github.com/istio/istio/pull/50006, some early code